---
title: "Protocol for metagenomic analyses"
author: "Karine Villeneuve"
date: "`r Sys.Date()`"
output: 
  html_document:
    fig_caption: yes
    highlight: textmate
    number_sections: yes
    theme: cosmo
    toc: yes
    toc_float: yes
  html_notebook: 
    toc: yes
geometry: margin=1cm
fontsize: 11pt
---

# Pre-Assembly 

## Unzip 
In the folder with your raw `fastq.gz`
```{bash, highlight=TRUE, eval=FALSE}
gunzip *.gz
```

## Interleaving 
I used the script <font color='green'>interleave_fastq.py</font>. You can also copy the script from the folder Script Métagénomique of our Teams.

Make the script executable and run the script using python 
```{bash, highlight=TRUE, eval=FALSE}
chmod +x interleave.py
python3 interleave.py 
```

Looping if you have multiple files 
```{bash, highlight=TRUE, eval=FALSE}
for R1 in *_R1_001.fastq ; do python3 interleave.py $R1 "${R1/R1/R2}"  > $R1.interleave.fastq ; done  
```

I am looking for a better way to rename the interleave file from the loop. But this will do for now. 

View the top of the new interleaved files to make sure R1 and R2 one after the others 
```{bash, highlight=TRUE, eval=FALSE}
grep @M interleaved.fastq | head
```
*grep (search) @M (symbol or letter) nameofdocument | (pipeline) head (show me just the first lines)* 

Move all the interleave files to a new folder called `interleave`

## SICKLE 

Running sickle with a bash script 

```
#!/bin/bash

for i in *.fastq

do
        sickle pe -c $i -t sanger -m $i.trim.fastq -s $i.singles.fastq
done
```
sickle pe (paired end) -c (inputfile) -t sanger (from illumina) -m (outputfilename) -s (exclutedreadsfilename)

This generates two files for a each fastq you input : `.trim.fastq` and `.singles.fastq`. We only need the `.trim.fastq` and therefor you can move all the others to another folder. 

## Quality check with Fastqc 

You can run FastQC in a non-interactive mode where you specify the files you want to process on the command line and FastQC will generate an HTML report for
each file without launching a user interface.  This would allow FastQC to be run as part of an analysis pipeline. To run non-interactively you simply have to specify a list of files to process on the commandline. You can specify as many files to process in a single run as you like.  

Run fastqc on the output (trimmed) file and the non-trimmed file

**VERY LONG SHOULD USE NANO** 
```
fastqc *.fastq --outdir=/home/kvilleneuve/Metagenomic_analyses/interleaved/fastqc
``` 
You must transfer the HTML files from the fastqc directory to your computer in order to view them in your a browser. 

## Fastq to Fasta 

Gzip the combined_trimmed.fastq file using nohup (very long) 
```
!#/bin/bash
gzip *.fastq
```
Convert the file from fastq to fasta
```
seqtk seq -a MF_combined_trimmed.gz > MF.fa
```





