---
title: "Setting up Titan Server"
author: "Karine Villeneuve"
date: "`r Sys.Date()`"
output: 
  html_document:
    fig_caption: yes
    highlight: textmate
    number_sections: yes
    theme: cosmo
    toc: yes
    toc_float: yes
  pdf_document:
    fig_caption: yes
    highlight: zenburn
    toc: yes
    toc_depth: 4
  word_document: default
  html_notebook: 
    toc: yes
geometry: margin=1cm
fontsize: 11pt
---
# Programs to download

The first thing I did was open this [protocol]() I wrote while I was in Texas and identify which programs and scripts were used.

## Python and anaconda /conda 

Python3 was already installed on the server. I did not need to install it. 


Anaconda 
```{bash, eval=FALSE}
wget https://repo.anaconda.com/archive/Anaconda3-2020.11-Linux-x86_64.sh -O ~/anaconda.sh
bash ~/anaconda.sh -p $HOME/anaconda
conda update --alls
conda config --set auto_activate_base false # If you'd prefer that conda's base environment not be activated on startup, set the auto_activate_base parameter to false: 
```
For some odd reason, on logging to the shell the server runs `.profile`  and not the `.bashrc`, therefor, we need to copy the PATH to python from the `.bashrc` shell to the `.profile` so it runs when logging in. 

## Sickle 

[Github](https://github.com/najoshi/sickle)

Sickle is a tool that uses sliding windows along with quality and length thresholds to determine when quality is sufficiently low to trim the 3’-end of reads and also determines when the quality is sufficiently high enough to trim the 5’-end of reads.

Dependencies : 
- [GCC](https://gcc.gnu.org/git.html)
- [Zlip](https://www.systutorials.com/how-to-install-the-zlib-library-in-ubuntu/) `sudo apt install zlib1g`

After installing these programs there was a folder called `sickle`. I moved into the folder and called the function `make` which prompted the installation of sickle. 


## FastQC

[Github](https://github.com/s-andrews/FastQC) Project [web page](http://www.bioinformatics.babraham.ac.uk/projects/fastqc/) for downloads 

FastQC is a program designed to spot potential problems in high throughput sequencing data sets. It runs a set of analyses on one or more raw sequence files in fastq or bam format and produces a report which summarizes the results. FastQC is a java application.  In order to run it needs your system to have a suitable
Java Runtime Environment (JRE) installed. You can check whether java is installed by opening the 'cmd' program on windows, or any shell on linux and typing: `java -version`.
If java isn't installed then you can add it by doing: `sudo apt install default-jre` 

Download the Fastqc program from the FastQC A quality control application for high throughput sequence 
data section (https://www.bioinformatics.babraham.ac.uk/projects/fastqc/)
Instructions for installation and setup (https://raw.githubusercontent.com/s-andrews/FastQC/master/INSTALL.txt)

```
git clone https://github.com/s-andrews/FastQC.git
sudo apt install fastqc 
```

## Fastq to fasta with seqtk

[Githup](https://github.com/lh3/seqtk)
To download:
```{bash, highlight=TRUE, eval=FALSE}
git clone https://github.com/lh3/seqtk.git;
cd seqtk; make
sudo apt install seqtk
```


## Assembly programs 

**IDBA**

[Github](https://github.com/loneknightpy/idba)

```
git clone https://github.com/loneknightpy/idba.git
sudo apt install idba
```

**Megahit**

[Github](https://github.com/voutcn/megahit)

```
conda install -c bioconda megahit
```

**metaSPAdes**


```
sudo apt install spades
```

## Mapping

**VE**

[Github](https://github.com/imrambo/genome_mapping)

```
git clone https://github.com/imrambo/genome_mapping.git
```

## Bin quality

**CheckM**

[Github](https://github.com/Ecogenomics/CheckM/wiki/Installation#how-to-install-checkm)

Must have these programs installed in the same path:

-HMMER
http://hmmer.org/

-prodigal

[Github](https://github.com/hyattpd/Prodigal)
```
sudo apt install prodigal
```

-pplacer
https://matsen.fhcrc.org/pplacer/

```
pip3 install numpy (or sudo apt install python3-pip)
pip3 install matplotlib
pip3 install pysam
pip3 install checkm-genome
```

## Bin Cleaning

**MM Genome**
MM Genome: http://madsalbertsen.github.io/multi-metagenome/docs/overview.html

```
wget https://github.com/MadsAlbertsen/multi-metagenome/tarball/master
tar -xzf master
```


## Phylogenetic tree

## Seqkit 

Download conda
Verify the latest release on the [website](https://repo.anaconda.com/archive/) and change the path is necessary. 

install 
```
conda install -c bioconda seqkit
```

**Phylosift**

[Github](https://github.com/gjospin/PhyloSift)

Download the tarball archive from

   http://edhar.genomecenter.ucdavis.edu/~koadman/phylosift/phylosift_latest.tar.bz2

Navigate to the directory where you downloaded the software and run :

```
tar -xvf phylosift_latest.tar.bz2
```

PhyloSift is ready for use (Refer to the Usage section) but may require dependancies. 

To see needed dependacies run from phylosift directory:
```
bin/phylosift
```

It may show this type of error message:

```
you may need to install the File::NFSLock module
```
Which means you need to run the following type of command (as `sudo` if needed):
```
cpanm File::NFSLock
```
where File::NFSLock is the name of the module required.

repeate till there are no more dependancies required.

if you do not have cpanm run:
```
cpan App::cpanminus
```
which installs it,

if you don't have cpan, install it.

```
sudo apt-get install git
```
To check out the PhyloSift development repository, run the following
command :
```
git clone https://github.com/rvosa/bio-phylo.git
perl Makefile.PL
make
make test # Optional, runs unit tests, which should pass
make install
```
```
git clone git://github.com/gjospin/PhyloSift.git
```

It may also be necessary to install several support packages,
including BioPerl, Bio::Phylo, and others. BioPerl offers bioperl-live from github,
we suggest installing other dependencies using cpan.

## Metabolic Pathway

**Metabolic**

[Github](https://github.com/AnantharamanLab/METABOLIC#install_instructions)

```
git clone https://github.com/AnantharamanLab/METABOLIC.git
```
In the PATH, these dependencies should be installed:

1. Perl (https://www.perl.org/)
2. HMMER (http://hmmer.org/)
3. Prodigal (https://github.com/hyattpd/Prodigal)
4. Sambamba (https://github.com/biod/sambamba)
5. BAMtools (https://github.com/pezmaster31/bamtools)
6. CoverM (https://github.com/wwood/CoverM)
7. R (https://www.r-project.org/)
8. Diamond (https://github.com/bbuchfink/diamond)
9. Samtools (https://www.htslib.org/)
10. Bowtie2 (https://github.com/BenLangmead/bowtie2)
11. Gtdb-tk (https://ecogenomics.github.io/GTDBTk/)

Perl modules
```
sudo apt install hmmer
sudo cpanm Data::Dumper
sudo cpanm POSIX
sudo cpanm Getopt::Long
sudo cpanm Statistics::Descriptive
sudo cpanm Array::Split
sudo cpanm Bio::SeqIO
sudo cpanm Bio::Perl
sudo cpanm Bio::Tools::CodonTable
sudo cpanm Carp
sudo cpanm File::Spec
sudo cpanm File::Basename
sudo cpanm Parallel::ForkManager
```

R packages
```
install.packages("versions")
library(versions)
install.versions("diagram", "1.6.4")
install.versions("forcats", "0.5.0")
install.versions("digest", "0.6.25")
install.versions("htmltools", "0.4.0")
install.versions("rmarkdown", "2.1")
install.versions("reprex", "0.3.0")
install.versions("tidyverse", "1.3.0")
install.versions("ggthemres", "4.2.0")
install.versions("ggalluvial", "0.11.3")
install.versions("reshape2", "1.4.3")
install.versions("ggraph", "2.0.2")
install.versions("pdftools", "2.3")
install.versions("igraph", "1.2.5")
install.versions("tidygraph", "1.1.2")
install.versions("stringr", "1.4.0")
install.versions("plyr", "1.8.6")
install.versions("dplyr", "0.8.5")
install.versions("openxlsx", "4.1.4")
q()
```
Ater the dependencies are installed, do:
```
pip install gdown
sudo sh run_to_setup.sh

perl METABOLIC-G.pl -test true
perl METABOLIC-C.pl -test true
```

# Getting the different scripts 

1. Interleaving : <font color='green'>interleave_fastq.py</font> 

Script was modified from [GitHub gist](https://gist.github.com/ngcrawford/2232505) in order to work using Python3. Script is now available in Teams. 
  
  Interleaving merge the R1 and R2 reads together in one file
  
2. GC content of assembly :  <font color='blue'>lenght.gc </font> 

Script was scp from Midgard (Texas) to my computer and is now available in Teams. (scp -P 2022 karine@146.6.184.138:/home/scripts/length+GC.pl .)
