---
title: "Setting up Titan Server"
author: "Karine Villeneuve"
date: "`r Sys.Date()`"
output: 
  html_document:
    fig_caption: yes
    highlight: textmate
    number_sections: yes
    theme: cosmo
    toc: yes
    toc_float: yes
  pdf_document:
    fig_caption: yes
    highlight: zenburn
    toc: yes
    toc_depth: 4
  word_document: default
  html_notebook: 
    toc: yes
geometry: margin=1cm
fontsize: 11pt
---
# Programs to download

The first thing I did was open this [protocol]() I wrote while I was in Texas and identify which programs and scripts were used.

## Python 

Python3 was already installed on the server. I did not need to install it. 

## Sickle 

[Github](https://github.com/najoshi/sickle)

Sickle is a tool that uses sliding windows along with quality and length thresholds to determine when quality is sufficiently low to trim the 3’-end of reads and also determines when the quality is sufficiently high enough to trim the 5’-end of reads.

Dependencies : 
- [GCC](https://gcc.gnu.org/git.html)
- [Zlip](https://www.systutorials.com/how-to-install-the-zlib-library-in-ubuntu/) `sudo apt install zlib1g`

After installing these programs there was a folder called `sickle`. I moved into the folder and called the function `make` which prompted the installation of sickle. 


## FastQC

[Github](https://github.com/s-andrews/FastQC) Project [web page](http://www.bioinformatics.babraham.ac.uk/projects/fastqc/) for downloads 

FastQC is a program designed to spot potential problems in high throughput sequencing data sets. It runs a set of analyses on one or more raw sequence files in fastq or bam format and produces a report which summarizes the results. FastQC is a java application.  In order to run it needs your system to have a suitable
Java Runtime Environment (JRE) installed. You can check whether java is installed by opening the 'cmd' program on windows, or any shell on linux and typing: `java -version`.
If java isn't installed then you can add it by doing: `sudo apt install default-jre` 

Download the Fastqc program from the FastQC A quality control application for high throughput sequence 
data section (https://www.bioinformatics.babraham.ac.uk/projects/fastqc/)
Instructions for installation and setup (https://raw.githubusercontent.com/s-andrews/FastQC/master/INSTALL.txt)

```
git clone https://github.com/s-andrews/FastQC.git
sudo apt install fastqc 
```

## Fastq to fasta with seqtk

[Githup](https://github.com/lh3/seqtk)
To download:
```{bash, highlight=TRUE, eval=FALSE}
git clone https://github.com/lh3/seqtk.git;
cd seqtk; make
sudo apt install seqtk
```


## Assembly programs 

**IDBA**

[Github](https://github.com/loneknightpy/idba)

```
git clone https://github.com/loneknightpy/idba.git
sudo apt install idba
```



# Getting the different scripts 

1. Interleaving : <font color='green'>interleave_fastq.py</font> 

Script was modified from [GitHub gist](https://gist.github.com/ngcrawford/2232505) in order to work using Python3. Script is now available in Teams. 
  
  Interleaving merge the R1 and R2 reads together in one file
  
2. GC content of assembly :  <font color='blue'>lenght.gc </font> 

Script was scp from Midgard (Texas) to my computer and is now available in Teams. (scp -P 2022 karine@146.6.184.138:/home/scripts/length+GC.pl .)
